### ISSUE TITLE
Error: whisper_full_with_state: encoder_begin_callback returned false - aborting

### ISSUE THREAD

**Issue by RicardoEPRodrigues** [gtreshchev/RuntimeSpeechRecognizer#18]
I hope you can help me with this. I do not know the source of the issue, but this seems to only happen in my code when it runs in C++. I'm using Unreal 5.3 on Windows (for reference).

If I run the blueprint example it runs just fine. It will loop the pending audio indefinitely (does the `pending: 79680, num of samples per step: 80000` then starts over). No problems here.

But if I run a similar code in C++, it does one loop alright, but at the end of the second a warning, then an error shows up.

`Warning: Aborting whisper recognition due to stop request`
`Error: whisper_full_with_state: encoder_begin_callback returned false - aborting`

And it never recovers. Can you replicate the behavior? (I leave a snippet of code below.)

<details><summary>Snippet of Log</summary>
<p>

```
PCM Info:
Validity of PCM data in memory: Valid, number of PCM frames: 160, PCM data size: 320
LogRuntimeSpeechRecognizer: Pending audio data instead of enqueuing it since it is not enough to fill the step size (pending: 79680, num of samples per step: 80000)
LogRuntimeAudioImporter: Successfully added audio data to streaming sound wave.
Added audio info: SoundWave Basic Info:
Number of channels: 2, sample rate: 16000, duration: 0.010000
PCM Info:
Validity of PCM data in memory: Valid, number of PCM frames: 160, PCM data size: 320
LogRuntimeSpeechRecognizer: Enqueued audio data from the pending audio to the queue of the speech recognizer (num of samples: 79840)
LogRuntimeAudioImporter: Successfully added audio data to streaming sound wave.
Added audio info: SoundWave Basic Info:
Number of channels: 2, sample rate: 16000, duration: 0.010000
PCM Info:
Validity of PCM data in memory: Valid, number of PCM frames: 160, PCM data size: 320
LogRuntimeSpeechRecognizer: Warning: Aborting whisper recognition due to stop request
LogRuntimeSpeechRecognizer: Warning: Aborting whisper recognition due to stop request
LogRuntimeSpeechRecognizer: Error: whisper_full_with_state: encoder_begin_callback returned false - aborting
LogRuntimeSpeechRecognizer: Processed audio data with the size of 79840 samples to the whisper recognizer
LogRuntimeAudioImporter: Successfully added audio data to streaming sound wave.
Added audio info: SoundWave Basic Info:
Number of channels: 2, sample rate: 16000, duration: 0.010000
PCM Info:
Validity of PCM data in memory: Valid, number of PCM frames: 160, PCM data size: 320
LogRuntimeSpeechRecognizer: Error: Audio processing failed: The audio data could not be processed to the recognizer since the thread is stopped
LogTemp: Error: DreamiaCharacter: Error in Speech Recognition: The audio data could not be processed to the recognizer since the thread is stopped
LogRuntimeSpeechRecognizer: Error: Audio processing failed: The audio data could not be processed to the recognizer since the thread is stopped
LogTemp: Error: DreamiaCharacter: Error in Speech Recognition: The audio data could not be processed to the recognizer since the thread is stopped
LogRuntimeAudioImporter: Successfully added audio data to streaming sound wave.
Added audio info: SoundWave Basic Info:
Number of channels: 2, sample rate: 16000, duration: 0.010000
```

</p>
</details> 

<details><summary>Code Snippet</summary>
<p>

### Header
```cpp
UFUNCTION(BlueprintCallable, Category="Dreamia|SpeechRecognition")
void StartAudioSession(bool bIsMuted = false);

UFUNCTION(BlueprintCallable, Category="Dreamia|SpeechRecognition")
void StopAudioSession();

UFUNCTION(BlueprintCallable, Category="Dreamia|SpeechRecognition")
void SetAudioSessionMute(const bool bIsMuted);

/**
* Audio Input Device ID (aka which microphone to use).
*/
UPROPERTY(BlueprintReadWrite, EditAnywhere, Category="Dreamia|SpeechRecognition")
int DeviceId = 0;

UPROPERTY()
FOnSpeechRecognitionStartedDynamic OnStartSpeechRecognitionEvent;

// STT Section
UPROPERTY()
bool IsAudioSessionMuted = false;

UPROPERTY()
class USpeechRecognizer* SpeechRecognizer;

UPROPERTY()
class UCapturableSoundWave* CapturableSoundWave;

UFUNCTION()
void OnRecognitionFinished();
UFUNCTION()
void OnRecognitionError(const FString& ShortErrorMessage, const FString& LongErrorMessage);
UFUNCTION()
void OnRecognizedTextSegment(const FString& RecognizedWords);

UFUNCTION()
void OnStartSpeechRecognition(bool bSucceeded);

UFUNCTION()
void OnPopulateAudioData(const TArray<float>& PopulatedAudioData);
```

### Code
```cpp
void UDreamiaCharacterComponent::StartAudioSession(const bool bIsMuted)
{
	SpeechRecognizer = USpeechRecognizer::CreateSpeechRecognizer();
	SpeechRecognizer->SetLanguage(ESpeechRecognizerLanguage::En);
	SpeechRecognizer->SetNumOfThreads(8);
	SpeechRecognizer->SetStepSize(1000);
	SpeechRecognizer->OnRecognitionFinished.AddUniqueDynamic(this, &UDreamiaCharacterComponent::OnRecognitionFinished);
	SpeechRecognizer->OnRecognitionError.AddUniqueDynamic(this, &UDreamiaCharacterComponent::OnRecognitionError);
	SpeechRecognizer->OnRecognizedTextSegment.AddUniqueDynamic(
		this, &UDreamiaCharacterComponent::OnRecognizedTextSegment);
	SpeechRecognizer->SetStreamingDefaults();

	OnStartSpeechRecognitionEvent.Clear();
	OnStartSpeechRecognitionEvent.BindDynamic(this, &UDreamiaCharacterComponent::OnStartSpeechRecognition);
	SpeechRecognizer->StartSpeechRecognition(OnStartSpeechRecognitionEvent);
}

void UDreamiaCharacterComponent::OnRecognitionError(const FString& ShortErrorMessage, const FString& LongErrorMessage)
{
	UE_LOG(LogTemp, Error, TEXT("%s: Error in Speech Recognition: %s"), *(GetNameSafe(this)), *LongErrorMessage);
}

void UDreamiaCharacterComponent::OnStartSpeechRecognition(bool bSucceeded)
{
	CapturableSoundWave = UCapturableSoundWave::CreateCapturableSoundWave();
	CapturableSoundWave->OnPopulateAudioData.AddUniqueDynamic(this, &UDreamiaCharacterComponent::OnPopulateAudioData);

	CapturableSoundWave->StartCapture(DeviceId);
	// CapturableSoundWave->ToggleMute(IsAudioSessionMuted);
}

void UDreamiaCharacterComponent::OnPopulateAudioData(const TArray<float>& PopulatedAudioData)
{
	if (!SpeechRecognizer)
	{
		// TODO LOG ERROR
		return;
	}
	if (!CapturableSoundWave)
	{
		// TODO LOG ERROR
		return;
	}

	SpeechRecognizer->ProcessAudioData(PopulatedAudioData, CapturableSoundWave->GetSampleRate(),
	                                   CapturableSoundWave->GetNumOfChannels(), false);
}

void UDreamiaCharacterComponent::OnRecognizedTextSegment(const FString& RecognizedWords)
{
	FString FilteredWords = RecognizedWords;

	// PROCESS TEXT
}

void UDreamiaCharacterComponent::OnRecognitionFinished()
{
	if (!SpeechRecognizer)
	{
		// TODO LOG ERROR
		return;
	}
	if (!CapturableSoundWave)
	{
		// TODO LOG ERROR
		SpeechRecognizer->StopSpeechRecognition();
		return;
	}

	if (CapturableSoundWave->IsCapturing())
	{
		SpeechRecognizer->StopSpeechRecognition();
	}
}

void UDreamiaCharacterComponent::StopAudioSession()
{
	if (CapturableSoundWave)
	{
		CapturableSoundWave->StopCapture();
	}
	if (SpeechRecognizer)
	{
		SpeechRecognizer->StopSpeechRecognition();
	}
}

void UDreamiaCharacterComponent::SetAudioSessionMute(const bool bIsMuted)
{
	IsAudioSessionMuted = bIsMuted;
	if (CapturableSoundWave && CapturableSoundWave->IsCapturing())
	{
		CapturableSoundWave->ToggleMute(IsAudioSessionMuted);
	}
}
```
</p>
</details>

**Comment by RicardoEPRodrigues** [comment]
I feel so dumb... like really really dumb!

The bug is on my end. It's in the `OnRecognitionFinished()`. Here:

```cpp
void UDreamiaCharacterComponent::OnRecognitionFinished()
{
	//...

	if (CapturableSoundWave->IsCapturing()) // <- It should be !CapturableSoundWave->IsCapturing()
	{
		SpeechRecognizer->StopSpeechRecognition();
	}
}
```

But it was very hard to debug because the problem was kinda obscured.

If you want to add a C++ example feel free to use this code (with the fix).

**Comment by gtreshchev** [comment]
Is it no longer relevant, were you able to resolve this?

Based on your logs, it seems that you tried to process the audio data before the speech recognizer thread started. But I haven't examined the code accurately, so I can look through it over the weekend if it's still an issue.

**Comment by gtreshchev** [comment]
> I feel so dumb... like really really dumb!
> 
> The bug is on my end. It's in the `OnRecognitionFinished()`. Here:
> 
> ```cpp
> void UDreamiaCharacterComponent::OnRecognitionFinished()
> {
> 	//...
> 
> 	if (CapturableSoundWave->IsCapturing()) // <- It should be !CapturableSoundWave->IsCapturing()
> 	{
> 		SpeechRecognizer->StopSpeechRecognition();
> 	}
> }
> ```
> 
> But it was very hard to debug because the problem was kinda obscured.
> 
> If you want to add a C++ example feel free to use this code (with the fix).

Ah, I see, great 👍
Yeah, thank you, I'll add a C++ example as well at some point :)

**Comment by RicardoEPRodrigues** [comment]
I modified the code to initialize everything before starting the audio session (now that the code is fixed). If you want I can share that example.

Thank you for the reply and amazing project!
